RPIP-46 spec update (#247)
rocket-pool/RPIPs: RPIPs/RPIP-46.md

1. The choices SHALL include retaining the state from Saturn 1
2. The choices MAY include [RPIP-45: RPL Burn](RPIP-45.md), [RPIP-50: RPL LP](RPIP-50.md), and novel options created for this vote
   1. The choices SHOULD specify a new share (eg, `buy_and_burn_share`) with an initial value
   2. The choices SHOULD specify a reduction of other shares (eg, `voter_share`) to balance the new share 
3. The selected mechanism MUST be implemented in Saturn 2
4. If the state from Saturn 1 is retained, the entirety of the ["Implementing the revenue share vote"](#implementing-the-revenue-share-vote) section below SHALL be deleted

## Specification taking effect with Saturn 2
### RPL rewards and inflation
1. Inflation settings SHALL be modified to retain inflation to the DAOs and eliminate inflation to NOs
   1. `rpl.inflation.interval.rate` SHALL be set to `1000040763630249500` (1.5% per year)
   2. Node Operators (`rocketClaimNode`) allocation SHALL be set to 0%
   3. pDAO (`rocketClaimDAO`) allocation SHALL be set to 95%
   4. oDAO (`rocketClaimTrustedNode`) allocation SHALL be set to 5%

> 2. There SHALL be no RPL rewards 

Reviewer: ```suggestion
2. There SHALL be no RPL issuance to NOs 
```

Reviewer: (voter_share, or other revenue sharing to RPL could generically be considered as "RPL rewards")

   4. `new_share -  node_operator_commission_share_council_adder`: this share of revenue is used to [PLACEHOLDER]
   5. `reth_commission` SHALL be defined as the sum of `node_operator_commission_share`, `voter_share`, and `new_share`
   6. `node_operator_commission_share`, `node_operator_commission_share_council_adder`, and `new_share`, SHALL be updateable by an address in the `allowlisted_controllers` array
2. `voter_share` SHALL no longer be a pDAO setting. the pDAO will not be able to vote changes to it and changes will rely on the method described below.
3. Updating `voter_share`:
   1. A permissionless function SHALL be available to update `voter_share`
   2. It MUST revert if it's been called within the last 45 days
   3. It MUST revert if the total RPL eligible to vote relative to the circulating supply of RPL is between `vote_eligible_target_min` and `vote_eligible_target_max` (inclusive)

>       1. RPL eligible includes both megapool staked RPL and legacy staked RPL

Reviewer: ```suggestion
      1. RPL eligible to vote includes both megapool staked RPL and legacy staked RPL
```

      1. `voter_share` is increased to `voter_share * (1+voter_share_relative_step)`
      2. `new_share` is decreased by the difference between the old and new `voter_share`
         1. If this would reduce `new_share` below 0%, the function call MUST revert 
   5. If >`vote_eligible_target_max` of total RPL is eligible to vote and the function succeeds:
      1. `voter_share` is decreased to `voter_share / (1+voter_share_relative_step)`
      2. `new_share` is increased by the difference between the old and new `voter_share`
   6. Because this involves _voters_ modifying `voter_share`, there is an acknowledged conflict of interest here. As a result, changing this method of "Updating `voter_share`" SHALL require a supermajority vote with at least 75% of the vote in support of any change.
4. The pDAO SHALL be able to update `voter_share_relative_step`, `vote_eligible_target_min`, and `vote_eligible_target_max` by vote; however, it SHALL require a supermajority vote with at least 75% of the vote in support of any change.
5. The initial settings SHALL be:
   1. `voter_share_relative_step`: 15%
   2. `vote_eligible_target_min`: 55%
   3. `vote_eligible_target_max`: 65%

> ## Optional heuristics [NEEDS UPDATE]

Reviewer: NEEDS UPDATE as in before general rework vote? Or before Saturn 1 goes live? Or something else

Author: As in "this PR is still a draft PR for a reason" 😛

## Specification

> ### UARS

Reviewer 2: ```suggestion
### Universal Adjustable Revenue Split
```

1. There SHALL be the following defined shares with settings: `node_operator_commission_share`, `node_operator_commission_share_council_adder`, `voter_share`
   1. `node_operator_commission_share + node_operator_commission_share_council_adder`: each NO receives this percentage of the commission from the borrowed ETH on validators they run. Unlike the remainder of the shares, this is _not_ a protocol revenue (ie, it is not socialized).

>    2. `voter_share -  node_operator_commission_share_council_adder`: each NO receives a share of revenue based on the vote-eligible RPL staked to their megapool. The overall voter share of revenue is based on the setting, and each NO receives a proportion of that based on `vote_eligible_RPL_in_their_megapool/total_vote_eligible_RPL_in_megapools`.

Reviewer 2: ```suggestion
   2. `voter_share - node_operator_commission_share_council_adder`: each NO receives a share of revenue based on the vote-eligible RPL staked to their megapool. The overall voter share of revenue is based on the setting, and each NO receives a proportion of that based on `vote_eligible_RPL_in_their_megapool / total_vote_eligible_RPL_in_megapools`.
```

Reviewer 2: match previous convention used for +

2. `reth_commission` SHALL be defined as the sum of `node_operator_commission_share`, and `voter_share`
3. `reth_share` SHALL be defined as `100% - reth_commission`
4. Distributions of revenue from borrowed ETH MUST respect the defined shares
   1. If shares change between claims, distributions MUST make an effort to account for the different values. For example, a distribution could use a duration-weighted average share. Approximations MAY be used where they significantly reduce complexity and/or costs.
   2. Legacy minipools are an exception and SHALL continue to support earlier distribution methodologies 

> 5. `node_operator_commission_share`, `node_operator_commission_share_council_adder`, and `voter_share`, SHALL be updateable by an address in the `allowlisted_controllers` array

Reviewer 2: ```suggestion
5. `node_operator_commission_share`, `node_operator_commission_share_council_adder`, and `voter_share`, SHALL be updateable by any address in the `allowlisted_controllers` array
```

   1. This functionality SHALL not be used without a separate pDAO vote to enable a controller and add it to the list
6. The `node_operator_commission_share_council_adder` setting SHALL only allow values where:

>    1. 0% ≤ `node_operator_commission_share_council_adder` ≤ `max_node_operator_commission_share_council_adder` 

Reviewer 2: ```suggestion
   1. `node_operator_commission_share_council_adder` ≤ `max_node_operator_commission_share_council_adder` 
   2. `node_operator_commission_share_council_adder` > 0%
```
Just more readable + less hard to conceptualize this way. I'm aware there is a small logical difference in that `max_node_operator_commission_share_council_adder` can now be less than zero, but this is irrelevant as the used value could then not be changed to an invalid value. 

If you want to keep it could add a third line, or just stick with the three-term expression.

Author: I find the original more readable. Not planning to change this, unless we hear back from other folks

Reviewer 2: You're both smart AF, am not surprised you're fine. I think others are more likely to struggle. I changed  0 <= X to X > 0 which should be equivalent?

>    2. `node_operator_commission_share_council_adder` ≤ `voter_share`

Reviewer 2: Oops, if you take above, also take this. 
```suggestion
   3. `node_operator_commission_share_council_adder` ≤ `voter_share`
```

7. The `node_operator_commission_share_council_adder` setting SHALL be controllable by the security council without requiring a delay

> 8. The security council SHOULD increment `node_operator_commission_share_council_adder` by 0.5% if the deposit pool is over half-full for the majority of a 2-week period with a constant `node_operator_commission_share + node_operator_commission_share_council_adder`

Reviewer 2: Is the team confirmed happy / accepting of this one? Last I saw @darrenlangley was against modifying it so frequently.

Author: ![image](https://github.com/user-attachments/assets/01f08da2-91b6-4e03-99cf-785bb2319dbf)
Commented on it when in PR. The comment is focused on complexity etc, but didn't seem significantly against. My take was "not a fan, but accepting", but potentially read too much into it.

Reviewer 3: I was commenting strictly from a saved dev effort perspective.

I am very worried about the security council being overreactive to short-term demand spikes. 

> The security council SHOULD increment `node_operator_commission_share_council_adder` by 0.5% if the deposit pool is over half-full for the majority of a 2-week period
This to me is crazy. 

I don't have a problem with delegating some of this power to save on pDAO voting and even being more responsive, we do not know how Saturn will react to demand. In my opinion before this is enabled we should have a separate discussion about how it is used - not bake it in now.

Reviewer 3: I should say that we are fine to include the functionality, as it allows us to adapt to uncertainty. I just have the concerns above.

    1. The security council SHALL NOT otherwise change `node_operator_commission_share_council_adder`
9. The initial pDAO settings SHALL be:
    1. `node_operator_commission_share`: 5%
    2. `node_operator_commission_share_council_adder`: 0%
    3. `voter_share`: 9%
    4. `max_node_operator_commission_share_council_adder`: 1%
    5. `allowlisted_controllers`: []

### RPL rewards
1. The RPL reward system specified in RPIP-30 SHALL continue
   1. All staked RPL SHALL be counted for this purpose (both "legacy staked RPL" and "megapool staked RPL" as defined in [RPIP-43](RPIP-43.md))

### Revenue share vote
Prior to the release of Saturn 1, a ranked-choice vote MUST be held to select a mechanism for revenue share to RPL

> 1. The choices SHALL include retaining the state from Saturn 1

Reviewer 2: ```suggestion
1. The choices SHALL include making no change prior to the release of Saturn 1.
```
'Retaining' imo is the wrong phrasing, given this is happening prior to release, you're talking about retaining something that hasn't come into active use yet.

Author: ```
1. The choices SHALL include the state from Saturn 1
...
4. If the state from Saturn 1 is chosen, ...
```
went really simple instead

> 2. The choices MAY include [RPIP-45: RPL Burn](RPIP-45.md), [RPIP-50: RPL LP](RPIP-50.md), and novel options created for this vote

Reviewer 2: ```suggestion
2. The choices SHOULD include [RPIP-45: RPL Burn](RPIP-45.md), [RPIP-50: RPL LP](RPIP-50.md)
3. The choices MAY include as-yet unwritten options in RPIP form. 
```
I'd argue the first should be SHOULD over MAY, we've talked about Burn and LP enough that I think people are expecting to vote on them unless there is a concrete reason we can explain not to, which imo is closer to SHOULD than MAY.

Reviewer 2: Maybe rewrite without the as-yet.

   1. The choices SHOULD specify a new share (eg, `buy_and_burn_share`) with an initial value
   2. The choices SHOULD specify a reduction of other shares (eg, `voter_share`) to balance the new share 
3. The selected mechanism MUST be implemented in Saturn 2

> 4. If the state from Saturn 1 is retained, the entirety of the ["Implementing the revenue share vote"](#implementing-the-revenue-share-vote) section below SHALL be deleted

Reviewer 2: ```suggestion
4. If no change from the Saturn 1 specification is required, the entirety of the ["Implementing the revenue share vote"](#implementing-the-revenue-share-vote) section below SHALL be deleted by the RPIP editors.
```

Reviewer 2: See above note on retaining something that doesn't exist at time-of-vote

## Specification taking effect with Saturn 2
### RPL rewards and inflation
1. Inflation settings SHALL be modified to retain inflation to the DAOs and eliminate inflation to NOs
   1. `rpl.inflation.interval.rate` SHALL be set to `1000040763630249500` (1.5% per year)
   2. Node Operators (`rocketClaimNode`) allocation SHALL be set to 0%
   3. pDAO (`rocketClaimDAO`) allocation SHALL be set to 95%
   4. oDAO (`rocketClaimTrustedNode`) allocation SHALL be set to 5%
2. There SHALL be no RPL issuance to NOs 

> ### Implementing the revenue share vote

Reviewer 2: ```suggestion
### Revenue Share Implementation
```

Reviewer 2: Oh, I see you've copied everything for this. Uhmmm, I feel like there must be a better way to do this that doesn't require keeping format / wording changes in sync between two not-quite identical things.

Author: It's actually possible to have one section do double duty. However, I kept it as two wholly separate copies to make it easier to remove a section post-vote, which is the plan.

Reviewer 2: Yeah, I understand. Don't love it, but whatever you think makes most sense is fine.

For this section, we'll be writing `new_share`. When the revenue share vote is passed, that will define the share's name and this section SHALL be updated.

> 1. The following updates SHALL be made in the [UARS](#uars) section of the specification above, with the placeholder description filled in:

Reviewer 2: Update header link if you take first suggestion

      1. `voter_share` is increased to `voter_share * (1+voter_share_relative_step)`
      2. `new_share` is decreased by the difference between the old and new `voter_share`
         1. If this would reduce `new_share` below 0%, the function call MUST revert 
   5. If >`vote_eligible_target_max` of total RPL is eligible to vote and the function succeeds:
      1. `voter_share` is decreased to `voter_share / (1+voter_share_relative_step)`
      2. `new_share` is increased by the difference between the old and new `voter_share`
   6. Because this involves _voters_ modifying `voter_share`, there is an acknowledged conflict of interest here. As a result, changing this method of "Updating `voter_share`" SHALL require a supermajority vote with at least 75% of the vote in support of any change.
4. The pDAO SHALL be able to update `voter_share_relative_step`, `vote_eligible_target_min`, and `vote_eligible_target_max` by vote; however, it SHALL require a supermajority vote with at least 75% of the vote in support of any change.
5. The initial settings SHALL be:
   1. `voter_share_relative_step`: 15%
   2. `vote_eligible_target_min`: 55%
   3. `vote_eligible_target_max`: 65%

> ## Optional heuristics

Reviewer 2: Given how long this is getting, I would like to advocate again for moving these sections to a supporting document and linking here. I guess alternatively you could use the click-to-expand functionality used here: <https://rpips.rocketpool.net/RPIPs/RPIP-54#outstanding-discrepancies>

Reviewer 2: If / once this become active, this could be undone.

Reviewer: I like the click-to-expand...

Author: Click to expand works for me

    4. `max_node_operator_commission_share_council_adder`: 1%
    5. `allowlisted_controllers`: []

### RPL issuance rewards

> - The `proposed_method_share` specified in [RPIP-30](RPIP-30.md) SHALL no longer be used; instead a new `proposed_method_share` SHALL be defined as follows:

Reviewer 4: I'm wondering if this is clearer if you just talk about the part of `proposed_method_share` that is supposed to change. Something like. "For <=10% borrowed ETH, `node_weight` is changed from 0 to [...], matching behavior between 10% and 15%".
 
Having it all spelled out is nice for the person implementing it and there is zero chance of any misunderstanding, but most readers care more about what's changing and you kind of have to look at RPIP-30 and compare to get that information.
